lib/jsdom/browser/resources/resource-loader.js

```diff
@@ -103,8 +82,56 @@ module.exports = class ResourceLoader {
 
       case "http":
       case "https": {
-        const requestOptions = this._getRequestOptions(options);
-        return request(urlString, requestOptions);
+        const { accept, cookieJar, referrer } = options;
+        const agents = agentFactory(this._proxy, this._strictSSL);
+        const requestOptions = {
+          followRedirects: true,
+          cookieJar,
+          agents,
+          headers: {
+            "User-Agent": this._userAgent,
+            "Accept-Language": "en",
+            "Accept-Encoding": "gzip",
+            "Accept": accept || "*/*"
+          }
+        };
+        if (referrer && !IS_BROWSER) {
+          requestOptions.headers.Referer = referrer;
+        }
+        const requestClient = new Request(urlString, requestOptions);
+        const promise = new Promise((resolve, reject) => {
+          const accumulated = [];
+          let accumulatedBytelength = 0;
+          requestClient.once("response", ({ statusCode }) => {
+            if (statusCode === 404) {
+              requestClient.abort();
+              reject(new Error("404 Not Found"));
+            } else if (statusCode >= 500) {
+              requestClient.abort();
+              reject(new Error("Error encountered"));
+            }
+          });
+          requestClient.on("data", chunk => {
+            accumulatedBytelength += chunk.length;
+            accumulated.push(chunk);
+          });
+          requestClient.on("end", () => {
+            try {
+              resolve(Buffer.concat(accumulated, accumulatedBytelength));
+            } catch (error) {
+              reject(error);
+            }
+          });
+          requestClient.on("error", reject);
+        });
+        requestClient.on("end", () => {
```

Comment:
Can you add comments explaining what parts of the code use response / href / getHeader? Eventually I'd like to eliminate those and use standard non-patched promises. (I know we need `abort()` pretty extensively, but I was surprised we need the others.)

Reply:
All of these three properties are used in the method `fromURL` in `lib/api.js`, and the omission of either one of them leads to 2650 tests crashing in this very function.

```javascript
  static fromURL(url, options = {}) {
    return Promise.resolve().then(() => {
      // Remove the hash while sending this through the research loader fetch().
      // It gets added back a few lines down when constructing the JSDOM object.
      const parsedURL = new URL(url);
      const originalHash = parsedURL.hash;
      parsedURL.hash = "";
      url = parsedURL.href;

      options = normalizeFromURLOptions(options);

      const resourceLoader = resourcesToResourceLoader(options.resources);
      const resourceLoaderForInitialRequest = resourceLoader.constructor === NoOpResourceLoader ?
        new ResourceLoader() :
        resourceLoader;

      const req = resourceLoaderForInitialRequest.fetch(url, {
        accept: "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
        cookieJar: options.cookieJar,
        referrer: options.referrer
      });

      return req.then(body => {
        const res = req.response;

        options = Object.assign(options, {
          url: req.href + originalHash,
          contentType: res.headers["content-type"],
          referrer: req.getHeader("referer")
        });

        return new JSDOM(body, options);
      });
    });
  }
```